sql/core/src/test/java/test/org/apache/spark/sql/JavaBeanWithArraySuite.java

```diff
@@ -0,0 +1,222 @@
+    }
+
+    @After
+    public void tearDown() {
+        spark.stop();
+        spark = null;
+    }
+
+    @Test
+    public void testBeanWithArrayFieldsDeserialization() {
+
+        StructType schema = createSchema();
+        Encoder<Record> encoder = Encoders.bean(Record.class);
+
+        Dataset<Record> dataset = spark
+                .read()
+                .format("json")
+                .schema(schema)
+                .load("src/test/resources/test-data/with-array-fields")
+                .as(encoder);
+
+        List<Record> records = dataset.collectAsList();
+
+        Assert.assertTrue(Util.equals(records, RECORDS));
+    }
+
+    private static StructType createSchema() {
+        StructField[] intervalFields = {
+                new StructField("startTime", DataTypes.LongType, true, Metadata.empty()),
+                new StructField("endTime", DataTypes.LongType, true, Metadata.empty())
+        };
+        DataType intervalType = new StructType(intervalFields);
+
+        DataType intervalsType = new ArrayType(intervalType, true);
+
+        DataType valuesType = new ArrayType(DataTypes.IntegerType, true);
+
+        StructField[] fields = {
+                new StructField("id", DataTypes.IntegerType, true, Metadata.empty()),
+                new StructField("intervals", intervalsType, true, Metadata.empty()),
+                new StructField("values", valuesType, true, Metadata.empty())
+        };
+        return new StructType(fields);
+    }
+
+    public static class Record {
+
+        private int id;
+        private List<Interval> intervals;
+        private List<Integer> values;
+
+        public Record() { }
+
+        Record(int id, List<Interval> intervals, List<Integer> values) {
+            this.id = id;
+            this.intervals = intervals;
+            this.values = values;
+        }
+
+        public int getId() {
+            return id;
+        }
+
+        public void setId(int id) {
+            this.id = id;
+        }
+
+        public List<Interval> getIntervals() {
+            return intervals;
+        }
+
+        public void setIntervals(List<Interval> intervals) {
+            this.intervals = intervals;
+        }
+
+        public List<Integer> getValues() {
+            return values;
+        }
+
+        public void setValues(List<Integer> values) {
+            this.values = values;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (!(obj instanceof Record)) return false;
+            Record other = (Record) obj;
+            return
+                    (other.id == this.id) &&
+                    Util.equals(other.intervals, this.intervals) &&
+                    Util.equals(other.values, this.values);
+        }
+
+        @Override
+        public String toString() {
+            return String.format("{ id: %d, intervals: %s }", id, intervals );
+        }
+    }
+
+    public static class Interval {
```

Comment:
This is duplicate to your another PR. Maybe we can consider put two tests in one Java file so we don't need to have two `Interval`.

Reply:
We can have this merged first and rebase another PR to have another test in the same file too.

Reply:
OK, sure.